### PR TITLE
Podcast playback update for iOS app native implementation

### PR DESCRIPTION
**Pull request by fdoxyz**
## What type of PR is this? 

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

In order for the native apps to make use of native frameworks to play/pause Podcasts we need to adapt the javascript behavior when browsing from the apps (iOS for now). The approach we're taking here is to reuse as much of existing HTML/JS Podcast controls as possible.

In order to try this out run your local server with this branch and browse from the `DEV-ios` repo (using the PR linked below)

Some of these changes include:
 - Supressing [web-based](https://www.w3schools.com/jsref/met_audio_play.asp) audio play/pause based on user agent identifier
 - Posting webkit messages for event management on the "native side"

## Related Tickets & Documents

 - [iOS repo issue](https://github.com/thepracticaldev/DEV-ios/issues/203)
 - [webkit messages implemented for haptic message delivery](https://github.com/thepracticaldev/dev.to/pull/1287)

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help
- [x] no, because this is still a WIP

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ ] no documentation needed
- [x] no, because **not sure** if needed

## Are there any post deployment tasks we need to perform?

This PR might require at least some sort of planning in advance of deployment to production. We wouldn't want to break (or at least minimize impact) on Podcast functionality for "non up-to-date" iOS app versions.

 ⚠️ **IMPORTANT** ⚠️ Please don't merge (even if we have review approvals) until we have defined a plan to release the iOS side of things, because this will currently break audio playback for the app 😇

## What gif best describes this PR or how it makes you feel?

hey arnold grooving with headphones

### CODE REVIEW

**Comment by fdoxyz** [comment]
Glad it's looking good! 

And indeed, there's actually a few other actions I need to tap into as well (i.e. mute, change playback speed, close podcast, etc). This will make sure all the web based controls work as expected. Hope to update the PR soon.

**Comment by fdoxyz** [comment]
Okay this PR now changes the way the Podcasts are being played. A brief summary of the changes up to this point:

- The `audio` HTML element no longer stores the current playback state (`currentTime`, `paused`, etc)
- The current playback state is stored in local storage and accessed through `currentAudioState()`
- Both (web & native) audio playback write to this object using `saveMediaState(currentState)`
- When streaming podcasts from a browser the `audio` HTML element handles the playback vs the native `AVPlayer` on iOS app
- The "media state" localstorage data is now bumped to a different key (`media_playback_state_v2`)

I'd love to hear any feedback from this refactor and of course would be great to test this some more because it's just the first working version of this new playback management

File: app/assets/javascripts/initializers/initializePodcastPlayback.js [hunk 1]
```diff
@@ -272,52 +341,128 @@ function initializePodcastPlayback() {
...
+
+  function getMediaState() {
+    var currentState = currentAudioState();
+    document.getElementById('audiocontent').innerHTML = currentState.html;
+    var audio = getById('audio');
+    if (audio == undefined) {
+      return;
+    }
+    if (!isNativeIOS()) {
+      audio.currentTime = currentState.currentTime || 0;
```
Comment (fdoxyz) [inline app/assets/javascripts/initializers/initializePodcastPlayback.js]: I believe this is the troublemaker that causes the playback glitch when browsing different pages

Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializePodcastPlayback.js]: I've tried to avoid setting `audio.currentTime` if the audio is currently playing (`!audio.paused`) but the audio HTML element is kind of behaving ina weird way (always shows that paused value as `false`). Still looking into this 🧐

Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializePodcastPlayback.js]: Resolved thanks to help from Ben  🎉. New commits to be shared soon

**Comment by fdoxyz** [comment]
Okay another update is here which includes the following:
- Makes sure media is only initialized once. Thanks for the help on this Ben!
- Changed the `getMediaState` function to `initializeMediaState` so it's easier to understand the purpose of the function
- Fixes some bugs (promise related in `playAudio()`, regarding initialization, etc)
- Reorganizing of functions in order to avoid complaints from codeclimate

Things are looking solid after some testing from my end for both web and iOS. Definitely eager to consider this PR ready to merge **after releasing the iOS app which handles the native playback***.

**Comment by fdoxyz** [comment]
Not a stupid question at all Josh! 

**TL;DR:** It seems quite tricky and not even sure if 100% possible/worth it.

**Actual brainstorm:**
I ran a quick search and couldn't find any tests that cover podcasts playing an episode properly. All existing tests so far focus on displaying the podcast episodes/dashboards or the podcast related workers being queued (it would actually be good if I'm wrong about this 😇)

After giving it some thought, a strategy that **might work** in order to add test coverage would look something like this:

1. Capybara tests on this repo where we "play" the podcasts (click on the records) from the website. This should spin the records.
2. Same test but using `DEV-Native-ios` User-Agent, which triggers the "native functionality". The records should spin just like before though.

Now it might seem like we're testing the same thing, but we can periodically check for the podcast time labels (i.e. `00:05 / 44:52`). In the first scenario the time should increment on its own because the website plays the audio and the listener will update this label.

On scenario number 2 the time label shouldn't change (because this is sent as a message from the native side). We could also manually (now role-playing as the native wrapper) send in the `time` event with some JS and check that the labels should reflect the time update. This means the mutation observer is receiving the message and updating the time labels correctly as the native wrapper would do when it plays the audio.

**The problem:**
To achieve this we would need a valid podcast created, published and also the worker to fetch valid episodes in order to actually play one (FactoryBot episode won't cut it bc it uses Fake data).

- Are we ok with streaming a real podcast episode multiple times on every single CI test run + contributors? 
- A workaround might be hosting a small file somewhere and streaming that instead but its URL needs to be within a Podcast episode fetched by the worker. Have we gone too far at this point for this test case? 

It seems like it's too complex to implement at the moment to be worth it if even possible IMHO. If there's another way I'd love to hear about it for sure 😄

Sorry for the super long comment... 
how can you read this its so long

**Comment by joshpuetz** [comment]
Thanks for the explanation @fdoxyz, that helps! I had a feeling testing it would be complicated. I feel like each Cabybara test streaming podcasts for ever test run is probably going to be the killer here! Either way: it feels like **way** too much for this PR!
